elephant/spike_train_generation.py

```diff
@@ -1060,5 +1064,91 @@ def compound_poisson_process(rate, A, t_stop, shift=None, t_start=0 * ms):
+    """
+    Returns a spike train whose spikes are a realization of a Poisson process
+    with the given rate and refractory period starting at time `t_start` and
+    stopping time `t_stop`.
+
+    All numerical values should be given as Quantities, e.g. 100*Hz.
+
+    Parameters
+    ----------
+
+    rate : Quantity scalar with dimension 1/time
+           The rate of the discharge.
+    refr_period : Quantity scalar with dimension time
+                  The time period the after one spike no other spike is
+                  emitted.
+                  Default: 2 ms
+    t_start : Quantity scalar with dimension time
+              The beginning of the spike train.
+              Default: 0 ms
+    t_stop : Quantity scalar with dimension time
+             The end of the spike train.
+             Default: 1000 ms
+    as_array : bool
+               If True, a NumPy array of sorted spikes is returned,
+               rather than a SpikeTrain object.
+               Default: False
+
+    Raises
+    ------
+    ValueError : If one of `rate`, `refr_period`, `t_start` and `t_stop`
+                 is not of type `pq.Quantity`.
+
+    Examples
+    --------
+        >>> from quantities import Hz, ms
+        >>> spikes = homogeneous_poisson_process_with_refr_period(
+            50*Hz, 3*ms, 0*ms, 1000*ms)
+        >>> spikes = homogeneous_poisson_process_with_refr_period(
+            20*Hz, 5*ms, 5000*ms, 10000*ms, as_array=True)
+
+    """
+    if not isinstance(t_start, Quantity) or not isinstance(t_stop, Quantity):
+        raise ValueError("t_start and t_stop must be of type pq.Quantity")
+    if not isinstance(refr_period, Quantity):
+        raise ValueError("refr_period must be of type pq.Quantity")
+    if not isinstance(rate, Quantity):
+        raise ValueError("rate must be of type pq.Quantity")
+
+    if t_stop.units != t_start.units:
+        t_stop = t_stop.rescale(t_start.units)
+
+    rate_mag = rate.rescale(1 / t_start.units).magnitude
+    refr_period_mag = refr_period.rescale(t_start.units).magnitude
+
+    if rate_mag >= 1. / refr_period_mag:
+        raise ValueError(
+            'The firing rate must be sufficiently smaller than the inverse of the refractory period, otherwise' +
+            f' the effective firing rate diverges. But now we have rate: {str(rate)} and' +
+            f' refractory period: {str(refr_period)}')
+
+    duration = (t_stop - t_start).magnitude
+    mean_spike_count = rate_mag * duration
+    spike_count = np.random.poisson(lam=mean_spike_count)
+
+    # Check that the number of spikes drawn from the Poisson distributions,
+    # can fit in the duration regarding the refractory period.
+    if spike_count >= 1. + duration / refr_period_mag:
+        spike_count = np.ceil(duration / refr_period_mag).astype(int)
+
+    eff_duration = duration - (spike_count - 1) * refr_period_mag
+
+    st = np.sort(np.random.random(spike_count)) * eff_duration
```

Comment:
I don't think that sorting is necessary here.
Otherwise, you implicitly make each interspike interval longer than the previous, which is not a feature of randomness.

Reply:
No, I just make each inter-spike interval positive...